tool/src/c2/ty.rs

```diff
@@ -52,61 +70,43 @@ pub type ResultType<'tcx> = (Option<&'tcx hir::OutType>, Option<&'tcx hir::OutTy
 
 /// Context for generating a particular type's header
 pub struct TyGenContext<'ccx, 'tcx, 'header> {
-    cx: &'ccx CContext<'tcx>,
-    header: &'header mut Header,
+    pub cx: &'ccx CContext<'tcx>,
+    pub decl_header: &'header mut Header,
```

Comment:
issue: we're only ever writing to one header at a time. I think TyGenContext should continue writing to one header, and we pass in two separate header objects in `gen_ty` above.

This also solves the need of carrying around `is_decl` (which i consider a very easy source of footguns) and the Result weirdness.

The idea is that this context is the context for generating a single header file relevant to a type (we probably can call it HeaderGenContext instead if we really want but i might start wanting to include type-specific stuff)

Reply:
I see `TyGenContext` as something that generates all the code for a specific type, not a specific header. This PR just formalizes that types have two types of header files. Let's discuss

Reply:
https://github.com/rust-diplomat/diplomat/issues/302